File: Source/Views/DisclosureView.m
/*
 
 Transcribbler, a Mac OS X text editor for audio/video transcription
 Copyright (C) 2013  Eli Bishop
 
 This program is free software; you can redistribute it and/or
 modify it under the terms of the GNU General Public License
 as published by the Free Software Foundation; either version 2
 of the License, or (at your option) any later version.
 
 This program is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.
 
 You should have received a copy of the GNU General Public License
 along with this program; if not, write to the Free Software
 Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 
 */

#import "DisclosureView.h"

#define kDefaultTitleHeight 26
#define kDefaultButtonLeft 2
#define kDefaultLabelLeft 20
#define kButtonBottomMargin 5
#define kLabelBottomMargin 2
#define kButtonSize 13
#define kLabelHeight 18


@implementation DisclosureView

- (id)initWithFrame:(NSRect)f
{
  self = [super initWithFrame:f];
  [super setAutoresizesSubviews:NO];
  inited = NO;
  enabled = YES;
  expanded = YES;
  indentContent = NO;
  fixedHeight = YES;
  titleHeight = kDefaultTitleHeight;
  preferredWidth = f.size.width;
  preferredHeight = f.size.height;
  return self;
}

- (void)dealloc
{
  [contentView release];
  [title release];
  [super dealloc];
}

- (void)setAutoresizesSubviews:(BOOL)flag
{
}

- (void)awakeFromNib
{
  if (inited) {
    return;
  }
  inited = YES;
  
  if (!title) {
    title = [[self toolTip] retain];
  }
  
  NSRect cvf = [self makeContentViewFrame];
  contentView = [[[NSView alloc] initWithFrame:cvf] retain];
  [contentView setAutoresizesSubviews:YES];
  for (NSView* v in [NSArray arrayWithArray:[self subviews]]) {
    [v removeFromSuperview];
    [contentView addSubview:v];
  }
  [contentView setAutoresizingMask:(NSViewWidthSizable | NSViewHeightSizable)];
  
  [super setAutoresizesSubviews:YES];
  [self setFrameSize:NSMakeSize(preferredWidth, preferredHeight + titleHeight)];

  if (expanded) {
    [self addSubview:contentView];
  }
  
  NSRect buttonFrame = NSMakeRect(kDefaultButtonLeft, preferredHeight + kButtonBottomMargin,
                                  kButtonSize, kButtonSize);
  disclosureButton = [[NSButton alloc] initWithFrame:buttonFrame];
  [disclosureButton setAutoresizingMask:(NSViewMaxXMargin | NSViewMinYMargin)];
  [disclosureButton setButtonType:NSOnOffButton];
  [disclosureButton setBezelStyle:NSDisclosureBezelStyle];
  [disclosureButton setTitle:@""];
  [disclosureButton setTarget:self];
  [disclosureButton setAction:@selector(toggle:)];
  [disclosureButton setState:(expanded ? NSOnState : NSOffState)];
  [disclosureButton setHidden:!enabled];
  [self addSubview:disclosureButton];
  
  NSRect labelFrame = NSMakeRect(kDefaultLabelLeft, preferredHeight + kLabelBottomMargin,
                                 preferredWidth - (kDefaultLabelLeft + 2), kLabelHeight);
  label = [[NSTextField alloc] initWithFrame:labelFrame];
  [label setAutoresizingMask:(NSViewWidthSizable | NSViewMinYMargin)];
  [label setEditable:NO];
  [label setStringValue:title];
  [label setBordered:NO];
  [label setDrawsBackground:NO];
  [self addSubview:label];
}

- (void)resizeSubviewsWithOldSize:(NSSize)oldSize
{
  [super resizeSubviewsWithOldSize:oldSize];
  if (expanded) {
    NSRect f = [self frame];
    preferredHeight = f.size.height - titleHeight;
    [contentView setFrame:[self makeContentViewFrame]];
  }
}

- (NSView*)contentView
{
  return contentView;
}

- (BOOL)isEnabled
{
  return enabled;
}

- (void)setEnabled:(BOOL)e
{
  if (enabled != e) {
    enabled = e;
    [disclosureButton setHidden:!e];
  }
}

- (BOOL)isExpanded
{
  return expanded;
}

- (void)setExpanded:(BOOL)e
{
  if (expanded != e) {
    expanded = e;
    [disclosureButton setState:(e ? NSOnState : NSOffState)];
    NSRect f = [self frame];
    
    if (!e) {
      preferredHeight = [contentView frame].size.height;
    }

    float newHeight = e ? (preferredHeight + titleHeight) : titleHeight;
    NSRect r = NSMakeRect(f.origin.x, f.origin.y - (newHeight - f.size.height),
                         f.size.width, newHeight);
    
    if (e) {
      [self setFrame:r];
      [contentView setFrame:[self makeContentViewFrame]];
      [self addSubview:contentView];
    }
    else {
      [contentView removeFromSuperview];
      [self setFrame:r];
    }

    [self setNeedsDisplay:YES];
  }
}

- (BOOL)isIndentContent
{
  return indentContent;
}

- (void)setIndentContent:(BOOL)flag
{
  if (flag != indentContent) {
    indentContent = flag;
    [contentView setFrame:[self makeContentViewFrame]];
  }
}

- (BOOL)fixedHeight
{
  return fixedHeight;
}

- (void)setFixedHeight:(BOOL)f
{
  fixedHeight = f;
}

- (float)preferredHeight
{
  return preferredHeight;
}

- (void)setPreferredHeight:(float)newHeight
{
  if (preferredHeight != newHeight) {
    preferredHeight = newHeight;
    
    if (expanded) {
      [self setFrameSize:NSMakeSize([self frame].size.width, titleHeight + newHeight)];
      [contentView setFrame:[self makeContentViewFrame]];
    }
  }
}

- (NSString*)title
{
  return title;
}

- (void)setTitle:(NSString*)t
{
  if (title != t) {
    [title release];
    title = [t retain];
    [label setStringValue:t];
  }
}

- (void)toggle:(id)sender
{
  if (sender != disclosureButton) {
    [disclosureButton setState:(([disclosureButton state] == NSOnState) ? NSOffState : NSOnState)];
  }
  [self setExpanded:([disclosureButton state] == NSOnState)];
}

- (NSSize)minimumSize
{
  return NSMakeSize(preferredWidth, expanded ? (preferredHeight + titleHeight) : titleHeight);
}

- (NSSize)maximumSize
{
  return expanded ? NSMakeSize(preferredWidth, fixedHeight ? (preferredHeight + titleHeight) : FLT_MAX) : [self minimumSize];
}

- (NSRect)makeContentViewFrame
{
  float x = indentContent ? kDefaultLabelLeft : 0;
  return NSMakeRect(x, 0, [self frame].size.width - x, preferredHeight);
}

@end

File: Source/Views/DisclosureView.swift
/*
 
 Transcribbler, a Mac OS X text editor for audio/video transcription
 Copyright (C) 2013-2017  Eli Bishop
 
 This program is free software; you can redistribute it and/or
 modify it under the terms of the GNU General Public License
 as published by the Free Software Foundation; either version 2
 of the License, or (at your option) any later version.
 
 This program is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.
 
 You should have received a copy of the GNU General Public License
 along with this program; if not, write to the Free Software
 Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 
 */

import Foundation

let kDefaultTitleHeight = CGFloat(26)
let kDefaultButtonLeft = CGFloat(2)
let kDefaultLabelLeft = CGFloat(20)
let kButtonBottomMargin = CGFloat(5)
let kLabelBottomMargin = CGFloat(2)
let kButtonSize = CGFloat(13)
let kLabelHeight = CGFloat(18)

@objc(DisclosureView)
public class DisclosureView: NSView, ViewSizeLimits {
  @IBOutlet private(set) var disclosureButton: NSButton!
  @IBOutlet private(set) var label: NSTextField!

  var contentView: NSView?
  var _title: String?
  var title: String? {
    get {
      return _title
    }
    set(t) {
      _title = t
      label?.stringValue = t ?? ""
    }
  }
  var inited: Bool = false
  var _enabled: Bool = true
  var enabled: Bool {
    get {
      return _enabled
    }
    set(e) {
      if (_enabled != e) {
        _enabled = e
        disclosureButton.isHidden = !e
      }
    }
  }
  var _expanded: Bool = true
  var expanded: Bool {
    get {
      return _expanded
    }
    set(e) {
      if (_expanded != e) {
        _expanded = e
        disclosureButton.state = e ? NSOnState : NSOffState
        if let cv = contentView {
          let f = self.frame
          if (!e) {
            preferredHeight = cv.frame.size.height
          }
          let newHeight = e ? (preferredHeight + titleHeight) : titleHeight
          let r = NSMakeRect(f.origin.x, f.origin.y - (newHeight - f.size.height),
                             f.size.width, newHeight)
          if (e) {
            self.frame = r
            resizeContentView()
            addSubview(cv)
          }
          else {
            cv.removeFromSuperview()
            self.frame = r
          }
          self.needsDisplay = true
        }
      }
    }
  }
  var fixedHeight: Bool = true
  var _indentContent: Bool = false
  var indentContent: Bool {
    get {
      return _indentContent
    }
    set(i) {
      if (_indentContent != i) {
        _indentContent = i
        resizeContentView()
      }
    }
  }
  var titleHeight: CGFloat = kDefaultTitleHeight
  var _preferredHeight: CGFloat = 0
  var preferredHeight: CGFloat {
    get {
      return _preferredHeight
    }
    set(h) {
      if (_preferredHeight != h) {
        _preferredHeight = h
        if (expanded) {
          self.setFrameSize(NSMakeSize(frame.size.width, titleHeight + h))
          resizeContentView()
        }
      }
    }
  }
  var preferredWidth: CGFloat = 0
  
  override init(frame f: NSRect) {
    super.init(frame: f)
    super.autoresizesSubviews = false
    preferredWidth = f.size.width
    _preferredHeight = f.size.height
  }
  
  required public init?(coder: NSCoder) {
    fatalError("init(coder:) has not been implemented")
  }
  
  override public func awakeFromNib() {
    if inited {
      return
    }
    inited = true
    
    if title == nil {
      title = toolTip
    }

    let cvf = makeContentViewFrame()
    let cv = NSView(frame: cvf)
    contentView = cv
    cv.autoresizesSubviews = true
    for v in subviews {
      v.removeFromSuperview()
      cv.addSubview(v)
    }
    cv.autoresizingMask = [NSAutoresizingMaskOptions.viewWidthSizable,
      NSAutoresizingMaskOptions.viewHeightSizable]
    
    super.autoresizesSubviews = true
    setFrameSize(NSMakeSize(preferredWidth, preferredHeight + titleHeight))
    
    if expanded {
      addSubview(cv)
    }
    
    let buttonFrame = NSMakeRect(kDefaultButtonLeft, preferredHeight + kButtonBottomMargin,
                                 kButtonSize, kButtonSize);
    disclosureButton = NSButton(frame: buttonFrame)
    disclosureButton.autoresizingMask = [NSAutoresizingMaskOptions.viewMaxXMargin,
                                         NSAutoresizingMaskOptions.viewMinYMargin]
    disclosureButton.setButtonType(NSOnOffButton)
    disclosureButton.bezelStyle = NSDisclosureBezelStyle
    disclosureButton.title = ""
    disclosureButton.target = self
    disclosureButton.action = #selector(toggle)
    disclosureButton.state = expanded ? NSOnState : NSOffState
    disclosureButton.isHidden = !enabled
    addSubview(disclosureButton)
    
    let labelFrame = NSMakeRect(kDefaultLabelLeft, preferredHeight + kLabelBottomMargin,
                                preferredWidth - (kDefaultLabelLeft + 2), kLabelHeight)
    label = NSTextField(frame: labelFrame)
    label.autoresizingMask = [NSAutoresizingMaskOptions.viewWidthSizable,
                              NSAutoresizingMaskOptions.viewMinYMargin]
    label.isEditable = false
    label.stringValue = title ?? ""
    label.isBordered = false
    label.drawsBackground = false
    addSubview(label)
  }

  override public func resizeSubviews(withOldSize oldSize: NSSize) {
    super.resizeSubviews(withOldSize: oldSize)
    if expanded {
      let f = frame
      preferredHeight = f.size.height - titleHeight
      resizeContentView()
    }
  }
  
  public func toggle(_ sender: Any) {
    if ((sender as! NSObject) != disclosureButton) {
      disclosureButton.state = (disclosureButton.state == NSOnState) ? NSOffState : NSOnState
    }
    expanded = (disclosureButton.state == NSOnState)
  }

  func minimumSize() -> NSSize {
    return NSMakeSize(preferredWidth, expanded ? (preferredHeight + titleHeight) : titleHeight)
  }
  
  func maximumSize() -> NSSize {
    return expanded ?
      NSMakeSize(preferredWidth, fixedHeight ? (preferredHeight + titleHeight) : CGFloat(FLT_MAX)) :
      minimumSize()
  }

  func makeContentViewFrame() -> NSRect {
    let x = indentContent ? kDefaultLabelLeft : 0
    return NSMakeRect(x, 0, frame.size.width - x, preferredHeight)
  }
  
  func resizeContentView() {
    if let cv = contentView {
      cv.frame = makeContentViewFrame()
    }
  }
}

